Wikipedia/UI-V5/WMFArticleListCollectionViewController.m

```diff
@@ -46,18 +46,33 @@ - (void)setDataSource:(SSArrayDataSource<WMFArticleListDataSource>* __nullable)d
         return;
     }
 
-    _dataSource.collectionView = nil;
+    _dataSource.collectionView     = nil;
+    self.collectionView.dataSource = nil;
 
     _dataSource = dataSource;
 
+    [_dataSource setSavedPageList:self.savedPages];
+
     if ([self isViewLoaded]) {
-        _dataSource.collectionView = self.collectionView;
+        if (_dataSource) {
+            _dataSource.collectionView = self.collectionView;
+            if ([_dataSource respondsToSelector:@selector(estimatedItemHeight)]) {
+                [self flowLayout].estimatedItemHeight = _dataSource.estimatedItemHeight;
+            }
```

Comment:
mainly because I went to let `SSDataSource` do its thing when the `collectionView` is set, which AFAICT means adding the items instead of reloading all the data (which also gives us a nicer transition).